### PR TITLE
Add more flexible errors

### LINKED ISSUE #76: ResourceHandler errors
### Description
I would like to have more flexibility in the errors returned in the resource handlers. 
Would you accept a PR with the following changes?

1. All ResourceHandler methods will return an `error` as the second return value.
2. Current errors will implement the `error` interface.
3. *Nil errors like `GetErrorNil` will be nil.
4. ScimError and its fields will be public
5. ScimType and the ScimType constants will be public.
6. The server will accept *ScimErrors, and the current errors by default.
7. The server will also accept errors implementing the interface `interface{ StatusCoder() int }` and will generate the proper ScimError with the detail message being the result of err.Error().

With a change like this I'll think I can make current code work as it's currently working and clean and make more flexible the resource handlers.

### PR DESCRIPTION
**Pull request by di-wu**
Closes #76

### CODE REVIEW

File: errors/error.go [hunk 1]
```diff
@@ -1,94 +1,234 @@
...
+// ScimErrorBadRequest returns an 400 SCIM error with the given message.
+func ScimErrorBadRequest(msg string) *ScimError {
+	return &ScimError{
+		Detail: msg,
+		Status: http.StatusBadRequest,
+	}
+}
+
+var (
+	// ScimErrorUniqueness returns an 409 SCIM error with a detailed message.
```
Comment (jdeflander) [inline errors/error.go:139]: Is there a reason for using pointer types here?

File: errors/error.go [hunk 2]
```diff
@@ -1,94 +1,237 @@
...
+// ScimErrorBadRequest returns an 400 SCIM error with the given message.
+func ScimErrorBadRequest(msg string) ScimError {
+	return ScimError{
+		Detail: msg,
+		Status: http.StatusBadRequest,
+	}
+}
+
+var (
+	// ScimErrorUniqueness returns an 409 SCIM error with a detailed message.
```
Comment (legosx) [inline errors/error.go:139]: You can create external vars for all of the error types
```suggestion
	// ScimErrorInvalidFilter returns an 409 SCIM error with a detailed message.
	ScimErrorInvalidFilter = ScimError{
		ScimType: ScimTypeInvalidFilter,
		Detail:   "The specified filter syntax was invalid (does not comply with Figure 1), or the specified attribute and filter comparison combination is not supported.",
		Status:   http.StatusBadRequest,
	}
	// ScimErrorTooMany returns an 409 SCIM error with a detailed message.
	ScimErrorTooMany = ScimError{
		ScimType: ScimTypeTooMany,
		Detail:   "The specified filter yields many more results than the server is willing to calculate or process.  For example, a filter such as \"(userName pr)\" by itself would return all entries with a \"userName\" and MAY not be acceptable to the service provider.",
		Status:   http.StatusBadRequest,
	}
	// ScimErrorUniqueness returns an 409 SCIM error with a detailed message.
```

File: errors/error.go [hunk 3]
```diff
@@ -1,94 +1,237 @@
...
+		Detail:   "The attempted modification is not compatible with the target attribute's mutability or current state.",
+		Status:   http.StatusBadRequest,
+	}
+	// ScimErrorInvalidSyntax returns an 400 SCIM error with a detailed message.
+	ScimErrorInvalidSyntax = ScimError{
+		ScimType: ScimTypeInvalidSyntax,
+		Detail:   "The request body message structure was invalid or did not conform to the request schema.",
+		Status:   http.StatusBadRequest,
+	}
+	// ScimErrorInvalidValue returns an 400 SCIM error with a detailed message.
```
Comment (legosx) [inline errors/error.go:157]:
```suggestion
	// ScimErrorInvalidPath returns an 400 SCIM error with a detailed message.
	ScimErrorInvalidPath = ScimError{
		ScimType: ScimTypeInvalidPath,
		Detail:   "The \"path\" attribute was invalid or malformed (see Figure 7).",
		Status:   http.StatusBadRequest,
	}
	// ScimErrorInvalidValue returns an 400 SCIM error with a detailed message.
```

File: errors/error.go [hunk 4]
```diff
@@ -1,94 +1,237 @@
...
+		Detail:   "A required value was missing, or the value specified was not compatible with the operation or attribute type, or resource schema.",
+		Status:   http.StatusBadRequest,
+	}
+	// ScimErrorNoTarget returns an 400 SCIM error with a detailed message.
+	ScimErrorNoTarget = ScimError{
+		ScimType: ScimTypeNoTarget,
+		Detail:   "The specified path did not yield an attribute or attribute value that could be operated on.",
+		Status:   http.StatusBadRequest,
+	}
+	// ScimErrorInternal returns an 500 SCIM error without a message.
```
Comment (legosx) [inline errors/error.go:169]:
```suggestion
	// ScimErrorInvalidVersion returns an 400 SCIM error with a detailed message.
	ScimErrorInvalidVersion = ScimError{
		ScimType: ScimTypeInvalidVersion,
		Detail:   "The specified SCIM protocol version is not supported (see Section 3.13).",
		Status:   http.StatusBadRequest,
	}
	// ScimErrorSensitive returns an 400 SCIM error with a detailed message.
	ScimErrorSensitive = ScimError{
		ScimType: ScimTypeSensitive,
		Detail:   "The specified request cannot be completed, due to the passing of sensitive (e.g., personal) information in a request URI.  For example, personal information SHALL NOT be transmitted over request URIs.  See Section 7.5.2.",
		Status:   http.StatusBadRequest,
	}
	// ScimErrorInternal returns an 500 SCIM error without a message.
```

**Comment by di-wu** [comment]
@legosx thank you for the suggestions! :1st_place_medal:

**Pull request merged**